docs/conditions.rst

```diff
@@ -249,3 +251,172 @@ do not need actual record by are only looking to traverse::
+	$compoundCondition3 = CompoundCondition::mergeAnd($condition5, $compoundCondition2);
+
+   // $compoundCondition is created using OR as junction and $compoundCondition1 and $compoundCondition3 as nested conditions
+	$compoundCondition = CompoundCondition::mergeOr($compoundCondition1, $compoundCondition3);
+	
+	
+CompoundCondition is an independent object not related to any model. Applying scope to model is using the Model::scope()->add($condition) method::
+
+	$contact->scope()->add($condition); // adding condition to a model
+	$contact->scope()->add($conditionXYZ); // adding more conditions
+	
+.. php:method:: __construct($nestedConditions = [], $junction = CompoundCondition::AND);
+
+Creates a CompoundCondition object from an array::
+
+	// below will create 2 conditions and nest them in a compound conditions with AND junction
+	$compoundCondition1 = new CompoundCondition([
+		['name', 'like', 'ABC%'],
+		['country', 'US']
+	]);
+	
+.. php:method:: negate();
+
+Negate method has behind the full map of conditions so any condition object can be negated, e.g negating '>=' results in '<', etc.
+For compound conditionss this method is using De Morgan's laws, e.g::
+
+	// using $compoundCondition1 defined above
+	// results in "(Name not like 'ABC%') or (Country does not equal 'US')"
+	$compoundCondition1->negate();
+
+.. php:method:: mergeAnd(AbstractCondition $conditionA, AbstractCondition $conditionB, $_ = null);
+
+Merge number of conditions using AND as junction. Returns the resulting CompoundCondition object.
+
+.. php:method:: mergeOr(AbstractCondition $conditionA, AbstractCondition $conditionB, $_ = null);
+
+Merge number of conditions using OR as junction. Returns the resulting CompoundCondition object.
+
+.. php:method:: simplify();
+
+Peels off single nested conditions. Useful for converting (((field = value))) to field = value.
+
+.. php:method:: clear();
+
+Clears the condition from nested conditions.
+
+.. php:method:: isOr();
+
+Checks if scope components are joined by OR
+
+.. php:method:: isAnd();
+
+Checks if scope components are joined by AND
+
+Conditions on Referenced Models
+-------------------------------
+
+Agile Data allows for adding conditions on related models for retrieval of type 'model has references where'.
+
+Setting conditions on references can be done utilizing the Model::refLink method but there is a shorthand format 
+directly integrated with addCondition method using "/" to chain the reference names::
+
+	$contact->addCondition('company/country', 'US');
+	
+This will limit the $contact model to those whose company is in US.
+'company' is the name of the reference in $contact model and 'country' is a field in the referenced model.
+
+If a condition must be set directly on the existence or number of referenced records the special symbol "#" can be
+utilized to indicate the condition is on the number of records::
+
+	$contact->addCondition('company/tickets/#', '>', 3);
```

Comment:
why not `company/tickets/count(*)`? (I just want a feedback on this, `#` may have sense)

Reply:
Ideas are welcome. Spent a lot of time figuring out what to use

Reply:
`count(*)` might be a source of issues as `count( * )` or `count(false)` can be hard to detect or evalueate to a wrong results (if we optimize on it).

of course, we can test for `count\s*\(` regex and throw on any other non-canonical input

I like `count(*)` more as it is clear on the first sight what it does, however is may/will be confusing if other expressions are not supported (which is stronger)

or do we want to support any expressions that contains `(` like `tickets/sum(price)`? (I am slightly against it now because of added complexity)

Reply:
if we have a data like:

company
MSFT
MMM
MMM

tickets
MSFT 1
MMM 1
MMM 2

and `$contact->addCondition('company/tickets/#', '>', 1);`

this should match only tickets for `MMM`, right? do we have a test for this usecase?

Reply:
Not sure what you mean by the MSFT and MMM but based on your condition only companies that have more than 1 referenced tickets will be included in the scope.
So if I understand right ony company MMM will be in the scope.
Test case is [here](https://github.com/atk4/data/blob/371d2faa1fad1ca3501f3f32d5245bd3506ba188/tests/ScopeTest.php#L186)

Reply:
yes, test `#`, `users/#` (should be the same) and the example above, I simplified it, in reality, it will be more like:

company
id name
1 MSFT
2 MMM

tickets
id company_id
1 1
2 2
3 2

as `count(*)` aggregates, testing this is very important

Reply:
If we don't implement other methods like `sum(price)`, `avg(price)` etc. this way, then I would say that `#` is much better to use than `count(*)`. It's simpler to test and simpler to write.

Reply:
I agree - but https://github.com/atk4/data/pull/660#discussion_r458686184 must be definitely tested

Reply:
We was discussing this with @mvorisek for about hour and came to conclusion that:
- currently support only #
- in future if we implement things like sum(field), avg(field), min(), max() etc, then we can also implement count(*) as longer version of #. But for now - just # is enough.

Reply:
I find it good idea to implement aggregates as well.
Maybe we can then use aggregares `any()` and `none()`? `#` will be the only shortcut

I am working now on level reference testing.

Reply:
Yes that sounds interesting - but in future, not now.
Initial implementation should be as simple and clean as possible or it will become to complex in future when adding more and more features into it. We suffered with that in toolkit code before as you can see :)

Reply:
Yes, I will look into that separately. Target is to merge base scope functionality first then aggregates, placeholders ...

Reply:
Finished with fixing and introducing tests on multilevel referenced conditions